Studio RPC Infrastructure (#2254)
zmkfirmware/zmk: docs/docs/development/studio-rpc-protocol.md

---
title: Studio RPC Protocol
---

:::warning
ZMK Studio is still in active development, and the below information is for development purposes only. For up to date information, join the [ZMK Discord](https://zmk.dev/community/discord/invite) server and discuss in `#studio-development`.
:::

## Overview

> The ZMK Studio UI communicates with ZMK devices using a custom RPC protocol developed to be robust, reliable, while remaining simple and easy te extend with future enhancements.

Reviewer: ```suggestion
The ZMK Studio UI communicates with ZMK devices using a custom RPC protocol developed to be robust, reliable, while remaining simple and easy to extend with future enhancements.
```

> The protocol consist of [protocol buffer](https://protobuf.dev/programming-guides/proto3/) messages which are encoded/decoded using message framing, and the transmitted using an underlying transport. Two transports are currently implemented, a BLE transport using a custom GATT service, and a serial port transport, which usually is used with CDC ADM devices over USB.

Reviewer: ```suggestion
The protocol consists of [protocol buffer](https://protobuf.dev/programming-guides/proto3/) messages which are encoded/decoded using message framing, and then transmitted using an underlying transport. Two transports are currently implemented: a BLE transport using a custom GATT service and a serial port transport, which usually is used with CDC ADM devices over USB.
```

## Protobuf Messages

The messages for ZMK Studio are defined in a dedicated [zmk-studio-messages](https://github.com/zmkfirmware/zmk-studio-messages) repository. Fundamentally, the [`Request`](https://github.com/zmkfirmware/zmk-studio-messages/blob/main/proto/zmk/studio.proto#L11) message is used to send any requests from the Studio client to the ZMK device, and the [`Response`](https://github.com/zmkfirmware/zmk-studio-messages/blob/main/proto/zmk/studio.proto#L21) messages are sent from the ZMK device to the Studio client.

> Responses can either be [`RequestResponses`](https://github.com/zmkfirmware/zmk-studio-messages/blob/main/proto/zmk/studio.proto#L28) that are are sent in response to an incoming `Request`, or a [`Notification`](https://github.com/zmkfirmware/zmk-studio-messages/blob/main/proto/zmk/studio.proto#L38) which is sent at any point from the ZMK device to the Studio client to inform the client about state changes on the device, e.g. device is unlocked.

Reviewer: ```suggestion
Responses can either be [`RequestResponses`](https://github.com/zmkfirmware/zmk-studio-messages/blob/main/proto/zmk/studio.proto#L28) that are sent in response to an incoming `Request` or a [`Notification`](https://github.com/zmkfirmware/zmk-studio-messages/blob/main/proto/zmk/studio.proto#L38) which is sent at any point from the ZMK device to the Studio client to inform the client about state changes on the device, e.g. that the device is unlocked.
```

## Message Framing

Studio uses a simple framing protocol to easily identify the start and end of a given message, with basic escaping to allow for unrestricted content.

The following special bytes are used for the framing protocol:

- Start of Frame (SoF): `0xAB`
- Escape Byte (Esc): `0xAC`
- End of Frame (EoF): `0xAD`

> A message consists of a SoF byte, the payload, escaped as needed, followed by an EoF byte. Within the payload, and of the special encoding bytes will be escaped by being prefixed with an Esc byte.

Reviewer: ```suggestion
A message consists of a SoF byte, the payload, escaped as needed, followed by an EoF byte. Within the payload, any of the special encoding bytes will be escaped by being prefixed with an Esc byte.
```
